README.md
# Flutter - Custom MediaPicker

A Flutter based Reusable component which provides MediaPicker with custom properties which can be used in any flutter application.


## Table of contents
- [Flutter Support](#flutter-support)
- [Demo](#demo)
- [Features](#features)
- [Getting started](#getting-started)
- [Usage](#usage)
- [Methods](#methods)
- [Directive options](#directive-options)
- [Want to Contribute?](#want-to-contribute)
- [Need Help / Support?](#need-help)
- [Collection of Components](#collection-of-Components)
- [Changelog](#changelog)
- [License](#license)
- [Keywords](#Keywords)


## Flutter Support

Version - Flutter 1.17 (stable)

We have tested our program in above version, however you can use it in other versions as well


## Demo
[![](wli_mediapicker.gif)](https://github.com/weblineindia/Flutter-Custom-MediaPicker/blob/master/wli_mediapicker.gif)

------

## Features

* Customisable MediaPicker with image and video.

* Common widget can be used for pick image, pick multiple images, pick video based on Type.

## Getting started

* Download this sample project and import custom_media_picker.dart file in your project..
* Update Widgets UI based on your requirements.


## Usage

Setup process is described below to integrate in project.

### Methods

Add below pub dev libraries

    //Put this libraries
      image_picker: ^0.6.7+4
      multi_image_picker: ^4.7.10

    //fire below command to sync libraries
      pub get

Configure CustomMediaPicker Widget

    //Put CustomMediaPicker Widget
      CustomMediaPicker()

Pass required properties

    //add child widget as per UI requirement
    //add picker type which identify to pick image or video
    //resource callback will accept future void method with two param
      CustomMediaPicker(
            childWidget: Image.asset(
                         'assets/images/ic_img.png',
                         height: screenHeight * 0.25),
            pickerType: PickerType.IMAGE,
            resourceCallback: resourceCallback,
    )


Available PickerType

      //required parameter to identify pick image or video
        PickerType.IMAGE //pick image
        PickerType.VIDEO //pick video

Available ResultType

      ResultType.IMAGE_PATH  //File value
      ResultType.VIDEO_PATH  //File value
      ResultType.IMAGE_ARRAY //List of assets

Available PickSource

      PickSource.CAMERA  //open camera
      PickSource.GALLERY  //open gallery
      PickSource.CHOOSE //open bottomSheet and will allow user to choose b/w image and gallery

Callback function

    //required parameter with future void method
      Future<void> resourceCallback(ResultType resultType,resourceData) async {
        //result type will identify resourceData type
        //you can typecast resource data and use as per your requirement
      }


Set below properties based on your requirements

    //optional property to pick from camera or gallery
    //default value will be PickSource.CAMERA
      pickSource: PickSource.CHOOSE,

    //optional property to select multiple image
    //it accepts boolean value, default value will be false
      isMultipleImage: true,

    //optional property for set max image limit
    //this property will use only when isMultipleImage is true
    //it accepts int, default value will be 2
      maxImageCount: 10,

    //optional property for set max video duration
    //this property will use only when picker type is PickerType.VIDEO
    //it accepts duration object, default will be  300 sec
      maxVideoDuration: this.maxVideoDuration = const Duration(seconds: 800),

    //optional property for set camera option
    //this property will use only when picker source is PickSource.CAMERA
    //it will accept CameraType object and default will be CameraDevice.rear
      this.preferredCameraDevice = CameraDevice.rear,

------

## Directive options

### iOS

    //Add the following keys to your Info.plist file, located in <project root>/ios/Runner/Info.plist:
    - NSPhotoLibraryUsageDescription - describe why your app needs permission for the photo library.
      This is called Privacy - Photo Library Usage Description in the visual editor.
    - NSCameraUsageDescription - describe why your app needs access to the camera.
      This is called Privacy - Camera Usage Description in the visual editor.
    - NSMicrophoneUsageDescription - describe why your app needs access to the microphone,
      if you intend to record videos. This is called Privacy - Microphone Usage Description in the visual editor.

### Android

    //API 29+
    - Add android:requestLegacyExternalStorage="true" as an attribute to the <application> tag in AndroidManifest.xml.
      The attribute is false by default on apps targeting Android Q.

------

## Want to Contribute?

- Created something awesome, made this code better, added some functionality, or whatever (this is the hardest part).
- [Fork it](http://help.github.com/forking/).
- Create new branch to contribute your changes.
- Commit all your changes to your branch.
- Submit a [pull request](http://help.github.com/pull-requests/).

------

## Need Help?
If you need help then feel free to contact our [Flutter developers](https://www.weblineindia.com/flutter-cross-platform-mobile-app-development.html).

 ------

## Collection of Components
 We have built many other components and free resources for software development in various programming languages. Kindly click here to view our [Free Resources for Software Development.](https://www.weblineindia.com/software-development-resources.html)

------

## Changelog
Detailed changes for each release are documented in [CHANGELOG](./CHANGELOG).

## License

[MIT](LICENSE)

[mit]: https://github.com/weblineindia/Flutter-Custom-MediaPicker/blob/master/LICENSE

## Keywords
Flutter Media Picker, Image Picker, Video Picker, Multiple Image Selection, Media Picker Demo, Weblineindia


CHANGELOG
1.0.0 (August 19, 2020)
Initial release - Sample source code for Flutter Reusable Media Picker Component.


test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_imagevideo_picker/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


lib/screens/maiscreen/main_screen.dart
import 'package:flutter/material.dart';
import 'package:flutter_imagevideo_picker/screens/detailscreen/detail_screen.dart';
import 'package:flutter_imagevideo_picker/screens/maiscreen/widget/custom_media_picker.dart';
import 'package:image_picker/image_picker.dart';
import 'package:multi_image_picker/multi_image_picker.dart';


class MainScreen extends StatefulWidget {
  @override
  _MainScreenState createState() => _MainScreenState();
}

class _MainScreenState extends State<MainScreen> {
  List<Asset> images = <Asset>[];
  final picker = ImagePicker();


  Future<void> resourceCallback(ResultType resultType,resourceData) async {
    Navigator.push(
      context,
      MaterialPageRoute(builder: (context) => DetailScreen(resultType, resourceData)),
    );
  }

  @override
  Widget build(BuildContext context) {
    final screenHeight = MediaQuery.of(context).size.height;
    final screenWidth = MediaQuery.of(context).size.width;
    return Scaffold(
      appBar: AppBar(
        title: const Text(
          'ImageVideo Picker',
          style: TextStyle(fontWeight: FontWeight.bold),
        ),
        centerTitle: true,
      ),
      body: Container(
        color: Colors.white,
        child: Column(
          children: [
            Expanded(
              child: Container(
                child: Column(
                  mainAxisAlignment: MainAxisAlignment.center,
                  children: [
                    CustomMediaPicker(
                      pickerType: PickerType.IMAGE,
                      pickSource: PickSource.CHOOSE,
                      isMultipleImage: true,
                      maxImageCount: 10,
                      resourceCallback: resourceCallback,
                      childWidget: Image.asset(
                        'assets/images/ic_img.png',
                        height: screenHeight * 0.25,
                      ),
                    ),
                    const SizedBox(height: 8.0),
                    const Text(
                      'Select Images',
                      style: TextStyle(fontSize: 20.0, color: Color(0xff00AFF0)),
                    ),
                    const SizedBox(height: 8.0),
                    Padding(
                      padding: EdgeInsets.symmetric(horizontal: screenWidth * 0.2),
                      child: const Text(
                        'Select single or multiple images from camera and gallery',
                        textAlign: TextAlign.center,
                      ),
                    ),
                  ],
                ),
              ),
            ),
            Container(
              height: 1,
              width: double.infinity,
              color: Colors.grey,
            ),
            Expanded(
              child: Column(
                mainAxisAlignment: MainAxisAlignment.center,
                children: [
                  CustomMediaPicker(
                    pickerType: PickerType.VIDEO,
                    pickSource: PickSource.CHOOSE,
                    resourceCallback: resourceCallback,
                    childWidget: Image.asset(
                      'assets/images/ic_video.png',
                      height: screenHeight * 0.25,
                    ),
                  ),
                  const SizedBox(height: 8.0),
                  const Text(
                    'Select Video',
                    style: TextStyle(fontSize: 20.0, color: Color(0xff00AFF0)),
                  ),
                  const SizedBox(height: 8.0),
                  Padding(
                    padding: EdgeInsets.symmetric(horizontal: screenWidth * 0.2),
                    child: const Text(
                      'Select video from camera and gallery',
                      textAlign: TextAlign.center,
                    ),
                  ),
                ],
              ),
            ),
          ],
        ),
      ),
    );
  }
}


lib/screens/detailscreen/detail_screen.dart
import 'dart:io';

import 'package:flutter/material.dart';
import 'package:flutter_imagevideo_picker/screens/detailscreen/widget/custom_image.dart';
import 'package:flutter_imagevideo_picker/screens/detailscreen/widget/custom_video.dart';
import 'package:flutter_imagevideo_picker/screens/detailscreen/widget/list_widget.dart';
import 'package:flutter_imagevideo_picker/screens/maiscreen/widget/custom_media_picker.dart';
import 'package:multi_image_picker/multi_image_picker.dart';
import 'package:video_player/video_player.dart';

// ignore: must_be_immutable
class DetailScreen extends StatelessWidget {
  const DetailScreen(this.dataType, this.data);

  // ignore: prefer_typing_uninitialized_variables
  final dataType, data;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
        appBar: AppBar(
          title: const Text(
            'ImageVideo Picker',
            style: TextStyle(fontWeight: FontWeight.bold),
          ),
          leading: IconButton(
            onPressed: () => Navigator.of(context).pop(),
            icon: const Icon(Icons.arrow_back),
          ),
          centerTitle: true,
        ),
        body: Container(
          width: double.infinity,
          color: Colors.white,
          child: Column(
            // ignore: prefer_const_literals_to_create_immutables
            children: [
              Padding(
                padding: const EdgeInsets.all(12.0),
                child: Text(
                  ((dataType as ResultType) == ResultType.VIDEO_PATH) ? 'Video' : 'Images',
                  style: const TextStyle(fontSize: 20.0, color: Color(0xff00AFF0)),
                ),
              ),
              Container(
                height: 1,
                width: double.infinity,
                color: Colors.grey,
              ),
              const SizedBox(
                height: 24,
              ),
              if ((dataType as ResultType) == ResultType.IMAGE_PATH) CustomImage(data as File),
              if ((dataType as ResultType) == ResultType.IMAGE_ARRAY)
                Expanded(
                  child: ListView(
                    scrollDirection: Axis.horizontal,
                    children: List.generate((data as List<Asset>).length, (index) {
                      return ListWidget((data as List<Asset>)[index]);
                    }),
                  ),
                ),
              if ((dataType as ResultType) == ResultType.VIDEO_PATH) CustomVideo(data as File)
            ],
          ),
        ));
  }
}


lib/screens/detailscreen/widget/list_widget.dart
import 'dart:io';

import 'package:flutter/material.dart';
import 'package:multi_image_picker/multi_image_picker.dart';

class ListWidget extends StatelessWidget {
  const ListWidget( this.data);

  // ignore: prefer_typing_uninitialized_variables
  final Asset data;

  // list image item for multiple image selection
  @override
  Widget build(BuildContext context) {
    return Stack(
      children: [
        Padding(
          padding: const EdgeInsets.all(8.0),
          child: AssetThumb(
            asset: data,
            width: 300,
            height: 180,
          ),
        ),
        Positioned(
          right: 0,
          top: 0,
          child: Image.asset(
            'assets/images/ic_close.png',
            height: 28,
            width: 28,
          ),
        ),
      ],
    );
  }
}


lib/screens/detailscreen/widget/custom_image.dart
import 'dart:io';

import 'package:flutter/material.dart';

class CustomImage extends StatelessWidget {
  const CustomImage( this.data);

  // ignore: prefer_typing_uninitialized_variables
  final File data;

  //custom image
  @override
  Widget build(BuildContext context) {
    final screenHeight = MediaQuery.of(context).size.height;
    final screenWidth = MediaQuery.of(context).size.width;
    return Stack(
      children: [
        Padding(
          padding: const EdgeInsets.all(8.0),
          child: Image.file(
            data,
            width: screenWidth * 0.9,
            height: screenHeight * 0.3,
            fit: BoxFit.cover,
          ),
        ),
        Positioned(
          right: 0,
          top: 0,
          child: Image.asset(
            'assets/images/ic_close.png',
            height: 28,
            width: 28,
          ),
        ),
      ],
    );
  }
}


lib/screens/detailscreen/widget/custom_video.dart
import 'dart:io';

import 'package:flutter/material.dart';
import 'package:video_player/video_player.dart';

class CustomVideo extends StatefulWidget {
  const CustomVideo(this.data);

  // ignore: prefer_typing_uninitialized_variables
  final File data;

  @override
  _CustomVideoState createState() => _CustomVideoState();
}

class _CustomVideoState extends State<CustomVideo> {
  VideoPlayerController _controller;

  @override
  void initState() {
    super.initState();
    //init controller
    _controller = VideoPlayerController.file(widget.data)
      ..initialize().then((_) {
        setState(() {});
      });
  }

  //Image thumbnail widget
  @override
  Widget build(BuildContext context) {
    final screenHeight = MediaQuery.of(context).size.height;
    final screenWidth = MediaQuery.of(context).size.width;
    return _controller.value.initialized
        ? Stack(
            children: [
              Padding(
                padding: const EdgeInsets.all(8.0),
                child: Container(
                  width: screenWidth * 0.9,
                  height: screenHeight * 0.3,
                  child: VideoPlayer(_controller),
                ),
              ),
              Positioned(
                right: 0,
                top: 0,
                child: Image.asset(
                  'assets/images/ic_close.png',
                  height: 28,
                  width: 28,
                ),
              ),
              Positioned(
                right: 0,
                top: 0,
                left: 0,
                bottom: 0,
                child: Image.asset(
                  'assets/images/ic_play.png',
                  height: 28,
                  width: 28,
                ),
              )
            ],
          )
        : Container();
  }
}


lib/screens/maiscreen/widget/custom_media_picker.dart
import 'dart:io';

import 'package:flutter/material.dart';
import 'package:image_picker/image_picker.dart';
import 'package:multi_image_picker/multi_image_picker.dart';

// ignore: must_be_immutable
class CustomMediaPicker extends StatelessWidget {
  CustomMediaPicker(
      {@required this.childWidget,
      @required this.pickerType,
      @required this.resourceCallback,
      this.pickSource = PickSource.CAMERA,
      this.isMultipleImage = false,
      this.maxImageCount = 2,
      this.maxVideoDuration = const Duration(seconds: 300),
      this.preferredCameraDevice = CameraDevice.rear});

  //customizable parameters
  final Widget childWidget;
  final PickerType pickerType;
  PickSource pickSource;
  final bool isMultipleImage;
  final int maxImageCount;
  final Function resourceCallback;
  final Duration maxVideoDuration;
  final CameraDevice preferredCameraDevice;

  //picker initialization
  var picker = ImagePicker();

  //method for pick images
  Future getImage() async {
    final imageData = await picker.getImage(
      source: pickSource == PickSource.CAMERA ? ImageSource.camera : ImageSource.gallery,
      preferredCameraDevice: preferredCameraDevice,
    );
    if (imageData != null) {
      resourceCallback(ResultType.IMAGE_PATH, File(imageData.path));
    }
  }

  //method for pick video
  Future getVideo() async {
    final videoData = await picker.getVideo(
        source: pickSource == PickSource.CAMERA ? ImageSource.camera : ImageSource.gallery,
        preferredCameraDevice: preferredCameraDevice,
        maxDuration: maxVideoDuration);
    if (videoData != null) {
      resourceCallback(ResultType.VIDEO_PATH, File(videoData.path));
    }
  }

  //method for pick multiple images
  Future<void> getMultipleImage() async {
    List<Asset> imageArray = <Asset>[];
    try {
      imageArray = await MultiImagePicker.pickImages(
        maxImages: maxImageCount,
        enableCamera: true,
        cupertinoOptions: const CupertinoOptions(takePhotoIcon: 'chat'),
        materialOptions: const MaterialOptions(
          actionBarColor: '#FDBC33',
          actionBarTitle: 'WeblineIndia',
          allViewTitle: 'All Photos',
          useDetailsView: false,
          selectCircleStrokeColor: '#000000',
        ),
      );
    } on Exception catch (e) {
      print(e.toString());
    }
    if (imageArray.isNotEmpty) resourceCallback(ResultType.IMAGE_ARRAY, imageArray);
  }

  //handle picker click
  void _handlePicker(BuildContext context) {
    if (isMultipleImage) {
      getMultipleImage();
    } else if (pickSource == PickSource.CHOOSE) {
      _settingModalBottomSheet(context);
    } else if (pickerType == PickerType.IMAGE) {
      getImage();
    } else if (pickerType == PickerType.VIDEO) {
      getVideo();
    }
  }

  //buid method for render ui
  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      onTap: () => _handlePicker(context),
      child: childWidget,
    );
  }

  //bottomsheet layout for choose b/w camera ad gallery
  void _settingModalBottomSheet(BuildContext context) {
    showModalBottomSheet(
        context: context,
        backgroundColor: const Color(0x00000000),
        builder: (BuildContext bc) {
          return Container(
            height: 240,
            margin: const EdgeInsets.symmetric(horizontal: 16.0),
            decoration: const BoxDecoration(
              borderRadius: BorderRadius.vertical(top: Radius.circular(24.0)),
              color: Colors.white,
            ),
            child: Column(
              // ignore: prefer_const_literals_to_create_immutables
              children: [
                const Padding(
                  padding: EdgeInsets.all(20.0),
                  child: Text(
                    'Choose From',
                    style: TextStyle(fontSize: 24.0, color: Color(0xff00AFF0), fontWeight: FontWeight.bold),
                  ),
                ),
                Container(
                  height: 1,
                  width: double.infinity,
                  color: Colors.grey,
                ),
                const SizedBox(
                  height: 44,
                ),
                Row(
                  mainAxisAlignment: MainAxisAlignment.spaceAround,
                  children: [
                    GestureDetector(
                      onTap: () {
                        Navigator.pop(context);
                        pickSource = PickSource.CAMERA;
                        if (pickerType == PickerType.IMAGE) {
                          getImage();
                        } else if (pickerType == PickerType.VIDEO) {
                          getVideo();
                        }
                      },
                      child: Column(
                        children: [
                          Image.asset('assets/images/ic_camera.png', height: 48),
                          const SizedBox(height: 8.0),
                          const Text('Camera', style: TextStyle(fontSize: 18.0))
                        ],
                      ),
                    ),
                    GestureDetector(
                      onTap: () {
                        Navigator.pop(context);
                        pickSource = PickSource.GALLERY;
                        if (pickerType == PickerType.IMAGE) {
                          getImage();
                        } else if (pickerType == PickerType.VIDEO) {
                          getVideo();
                        }
                      },
                      child: Column(
                        children: [
                          Image.asset('assets/images/ic_gallery.png', height: 48),
                          const SizedBox(height: 8.0),
                          const Text('Gallery', style: TextStyle(fontSize: 18.0))
                        ],
                      ),
                    ),
                  ],
                )
              ],
            ),
          );
        });
  }
}

//enum for pick image or video
enum PickerType { IMAGE, VIDEO }
//eum for pick from camera or gallery or choose from both
enum PickSource { CAMERA, GALLERY, CHOOSE }
//eum for result type
enum ResultType { IMAGE_PATH, VIDEO_PATH, IMAGE_ARRAY }
